src/fparser/two/Fortran2008.py

```diff
@@ -526,6 +679,55 @@ class Upper_Cobound(Base):  # R513
     subclass_names = ['Specification_Expr']
 
 
+class Do_Term_Action_Stmt(Do_Term_Action_Stmt_2003):  # R826
+    """
+    Fortran 2008 rule R826
+    do-term-action-stmt is action-stmt
+
+    Associated constraints are:
+
+    "C816 (R826) A do-term-action-stmt shall not be an arithmetic-if-stmt,
+          continue-stmt, cycle-stmt, end-function-stmt, end-mp-subprogram-stmt,
+          end-program-stmt, end-subroutine-stmt, error-stop-stmt, exit-stmt,
+          goto-stmt, return-stmt, or stop-stmt."
+    """
+    subclass_names = ['Action_Stmt_C816']
+
+
+class If_Stmt(If_Stmt_2003):  # R837
+    '''
+    Fortran 2008 rule R837
+    if-stmt is IF ( scalar-logical-expr ) action-stmt
+
+    The implementation of this rule does not add anything to the Fortran 2003
+    variant but reimplements the match method identical to Fortran 2003 as
+    otherwise the updated Fortran 2008 variant of `Action_Stmt_C828` (C802 in F2003)
+    would not be used.
+
+    Associated constraints are:
+
+    C828 (R837) The action-stmt in the if-stmt shall not be an end-function-stmt,
+          end-mp-subprogram-stmt, end-program-stmt, end-subroutine-stmt, or if-stmt.
+
+    '''
+    use_names = ['Scalar_Logical_Expr', 'Action_Stmt_C828']
+
+    @staticmethod
```

Comment:
I appreciate you've probably copied what has been done elsewhere but I think we could avoid code duplication by parameterising the method in the original class so that the two classes used by the `return` statement (`Scalar_Logical_Expr`, `Action_Stmt_C828` in this case) are class attributes. Then, all this class would have to do would be to set those attributes accordingly. As a carrot, that means you don't have to add a docstring to this method because it will go :-)

Reply:
That sounds like a good idea, will give it a try.

Reply:
That carrot was quite a mouthful to chew... ;-) (Unless I misunderstood your suggestion, of course)

To parametrize `match` I converted it to a `classmethod` and introduced an attribute `action_stmt_cls` which is then used as `cls.action_stmt_cls` in the return statement. In F2008 I can then only overwrite this attribute.

_But_: Using the classmethod from the parent means `Fortran2008.If_Stmt` no longer has a `'match'` entry in its `__dict__` - hence it is excluded when building the `Base.subclasses` dict in `parser.py`: https://github.com/stfc/fparser/blob/5b9a8467da393f3a88aaa172d5add776d9d7add2/src/fparser/two/parser.py#L226

This can be worked around by using `hasattr` in this check instead but I don't know if this might have other implications? (Tests are happy, though, and the possibly slight overhead probably doesn't matter as this is done only in the initial setup?).

Reply:
I've checked with the PSyclone test suite and with the NEMO code base and everything seems fine.